code/espurna/sensors/A02YYUSensor.h

```diff
@@ -0,0 +1,148 @@
+// DYP A02YYUx Ultrasonic Distance Sensor
+// Copyright (C) 2024 by tsouthiv @ gmail dot com
+// -----------------------------------------------------------------------------
+
+// This should be also compatible with other similar sensors like JSN-SR04T
+// or AJ-SR04M in auto UART mode.
+
+#if SENSOR_SUPPORT && A02YYU_SUPPORT
+
+#pragma once
+
+#include "Arduino.h"
+#include "BaseSensor.h"
+
+#define _RX_LEN     16
+#define _RX_MASK    (_RX_LEN - 1)
+
+class A02YYUSensor : public BaseSensor {
+
+    public:
+
+        void setPort(Stream* port) {
+            _serial = port;
+            _dirty = true;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        unsigned char id() const override {
+            return SENSOR_A02YYU_ID;
+        }
+
+        unsigned char count() const override {
+            return 1;
+        }
+
+        // Initialization method, must be idempotent
+        void begin() {
+            if (!_dirty) return;
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() const override {
+            return F("A02YYU");
+        }
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) const override {
+            return String(A02YYU_PORT, 10);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) const override {
+            if (index == 0) return MAGNITUDE_DISTANCE;
+            return MAGNITUDE_NONE;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) override {
+            if (index == 0) return _distance;
+            return 0;
+        }
+
+        // Loop-like method, call it in your main loop
+        void tick() override {
+            _read();
+        }
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        int _parse(uint8_t startIndex)
+        {
+            uint8_t i = startIndex;
+
+            if (_rx[i] != 0xff)
+                return -1;
+
+            int vh = _rx[(i + 1) & _RX_MASK];
+            int vl = _rx[(i + 2) & _RX_MASK];
+            int cs = _rx[(i + 3) & _RX_MASK];
+
+            if (cs != ((0xff + vh + vl) & 0xff))
+                return -2;
+
+            return (vh << 8) + vl;
+        }
+
+        void _read() {
+
+            int n;
+
+            _error = SENSOR_ERROR_OK;
+
+            // read all available bytes into rx circular buffer
+            while ((n = _serial->available()))
+            {    
+                for (; n > 0; n--) 
+                {
+                    char c = _serial->read();
```

Comment:
According my understanding of stream.readBytes(), it would block if there is not enough data and timeout (defined by setTimeout()).

The point here is to implement an asynchronous read/consume process through a circular buffer which is fed at a given input pace, while being parsed anytime there is enough data. The  per byte read facilitates the wrapping of the circular buffer. Would you suggest that it would have been better to implement it using for instance 2x readBytes() (one to fill until the end of the buffer and the second to fill the beginning on wrapping situation)?

Also the motivation to use such circular buffer is to offer some error resisience against erroneous or missing data. In case the parse() fails, it just drops the oldest received byte and try again starting with the next received byte ensuring that if any valid data is received, it would be properly parsed. This demonstrated pretty good result with a faulty sensor I had, this code was able to recover a few valid measurements here and there among randomly broken data.

Reply:
> Would you suggest that it would have been better to implement it using for instance 2x readBytes() (one to fill until the end of the buffer and the second to fill the beginning on wrapping situation)?

Something like that. We already have 'available' recorded, reading with its value would not block.

> Also the motivation to use such circular buffer is to offer some error resisience against erroneous or missing data. In case the parse() fails, it just drops the oldest received byte and try again starting with the next received byte ensuring that if any valid data is received, it would be properly parsed. This demonstrated pretty good result with a faulty sensor I had, this code was able to recover a few valid measurements here and there among randomly broken data.

I am just vary of having the reader tracking a lot of state, when essentially the algo is 'find the 0xff and read the next 3 bytes'

BTW are you sure `_error = OVERFLOW` below is valid? Shouldn't it scan the whole 'available' instead of dropping it midway?

Reply:
Following your comments, I revisited the approach and would like to suggest a simpler implementation of the processing loop. Rather than working with a circular buffer and asynchonous read/consume, the draft code below actually uses a linear frame buffer and parses it immediately as soon as there are enough data.

Let me know if this looks easier to review. If so, I'll finalize it, do some tests and validation before resubmitting it.

Thanks.

```cpp
#define _FRAME_LEN 4            // frame buffer length
uint8_t _frame[_FRAME_LEN];     // distance frame data buffer
int _free;                      // index of the first free byte in the frame buffer

void _read()
{
    int available = _serial->available();

    while (available) 
    {
        // read bytes from the device into the frame buffer
        int n = readBytes(_frame + _free, min(_FRAME_LEN - _free, available));

        available -= n;
        _free += n;
        
        if (_free == _FRAME_LEN) // the frame buffer is filled
        {
            int d = _parse_frame();
            
            if (d >= 0)
            {
                // distance successfully parsed
                _distance = d;      // update the distance
                _free = 0;          // empty the frame buffer
            } else {
                // out of sync or corrupted, shift the frame buffer by one byte
                for (int i = 0; i < _FRAME_LEN - 1; i++)
                    _frame[i] = _frame[i + 1];

                _free = _FRAME_LEN - 1;
            }
        }
    }
}
```

Reply:
It does look more streamlined, thanks.

Considering 'faulty sensor', should you need some validation checks to persist in the repo, I have also tried to fix local sensor builds w/ cmake
https://github.com/xoseperez/espurna/commit/2611b6294f5840d9ba7492d8588074c27d4dc44d#diff-b726c4b4ac60a04fa60bfd29302c3fb9aa412748579309e39ff1974c336a7080R1
code/test/unit/src/sensor/sensor.cpp *should* be able to pull the sensor header and call ->tick() then ->value(0) / ->error() reading from a provided Stream

Reply:
I've committed the simplified implementation and a few cleanups. It has run overnight successfully. So far, I believe the open items have been covered. Let me know if I missed anything.